tests_common/deflake.py

```diff
@@ -0,0 +1,341 @@
+    retrying them and marking them as "flaky" instead of "passing" if they
+    pass on the retry. After the test run, a GitHub issue is opened or updated
+    for each flaky test encountered.
+    """
+
+    name = "deflake"
+
+    def __init__(self: DeflakePlugin, config: Config) -> None:
+        super().__init__()
+        self.exceptions: dict[str, Exception] = {}
+        self.flaky_tests = 0
+        self.test_run = TestRun()
+
+        self.runner: type[runner] = config.pluginmanager.getplugin("runner")
+        self.test_type: str = config.getoption(TEST_TYPE_OPTION)
+
+        if token := os.getenv("GH_TOKEN"):
+            # Grab the current commit so we can generate absolute URLs
+            rev_parse = run(["git", "rev-parse", "HEAD"], text=True, stdout=PIPE)
+            self.github: GitHub | None = GitHub(token, rev_parse.stdout.strip())
+        else:
+            self.github = None
+
+    def pytest_sessionstart(self, session: Session) -> None:
+        # The session is the root node in pytest's collection tree
+        # Its path is the root directory that pytest looks in for tests, in
+        # our case it's the root of the repo, so we can use it to make
+        # relative paths that can be used to generate GitHub urls
+        self.test_run.root = session.path
+
+    def pytest_runtest_protocol(self, item: Item, nextitem: Item | None) -> bool:
+        # The main protocol for running a single test
+        # This function was adapted from pytest internals since we
+        # need to override it completely to retry a test
+
+        ihook = item.ihook
+
+        # Call the logstart hook, which typically prints the name of the tests
+        ihook.pytest_runtest_logstart(nodeid=item.nodeid, location=item.location)
+
+        # Delegate back to the pytest runner to run all the phases for this test
+        reports = self.runner.runtestprotocol(item, nextitem=nextitem)
+
+        # Retry if the test reports that it should retry (report.should_retry is
+        # set by us in pytest_runtest_makereport below)
+        if any(report.should_retry for report in reports):
+            self.runner.runtestprotocol(item, nextitem=nextitem)
+
+        # Close the log line for this test
+        ihook.pytest_runtest_logfinish(nodeid=item.nodeid, location=item.location)
+
+        # Just return something so the next hook implementation isn't called
+        return True
+
+    def pytest_runtest_logstart(self, nodeid: str):
+        # This is called once before each test regardless of the number of retries
+        self.test_run.tests[nodeid] = TestResult(nodeid)
+
+    @pytest.hookimpl(hookwrapper=True)
+    def pytest_runtest_makereport(self, item: Item, call: CallInfo) -> TestReport:
+        # Wrap the creation of the test report for this phase to override the
+        # status to "flaky" if it's a retry and it passed
+        previous_outcomes = item.stash.setdefault(PREVIOUS_OUTCOME_KEY, {})
+        report = (yield).get_result()
```

Comment:
Can you explain this bit of magic? I've never seen something like that 🤔

Reply:
generators in python can receive values after yielding, which is received as the result of the yield. here, pytest is sending an object of type `Result`, which has a `get_result()` method to get the actual return value of the hook being wrapped. 

the specific `report = (yield).get_result()` syntax is just a shorthand for
```python
outcome = yield
report = outcome.get_result()
```

More information on hook wrappers: https://docs.pytest.org/en/7.1.x/how-to/writing_hook_functions.html#hookwrapper-executing-around-other-hooks

Reply:
Thanks! Can we do
```
outcome: Result = yield  # yield nothing, receive result when generator resumes
report = outcome.get_result()
```
I think it will help our future selves

Reply:
done